### DISCUSSION TITLE
Enable collision avoidence

### DISCUSSION THREAD

**Discussion by Mardank** [USC-ACTLab/crazyswarm#644] [Q&A] [answered]
Hey all,

currently, I'm trying to set up crazyswarm on my system. We are using lighthouse V1 setup with the crazyflie(2.1). I have been able to set up and successfully fly with more than two drones. Now I want to test the onboard collision avoidance with two or more drones. Down below there is a little demo to test the onboard collision avoidance feature. Unfortunately, the collision avoidance didn't work in the experiments. However, in the simulation it works as intended.

Is there anything special I have to be aware of when using it?

Kind regards,
Markus

```
#!/usr/bin/env python

import numpy as np
from pycrazyswarm.crazyswarm_py import Crazyswarm 

Z = 0.5  # Takeoff altitude.

def main():
    swarm = Crazyswarm()
    timeHelper = swarm.timeHelper
    allcfs = swarm.allcfs
    cfs =  allcfs.crazyflies

    swarm.allcfs
    cf_one = cfs[0]
    cf_two = cfs[1]

    cf_one.enableCollisionAvoidance([cf_two], np.array([0.3, 0.3, 0.4]))
    cf_two.enableCollisionAvoidance([cf_one], np.array([0.3, 0.3, 0.4]))

    allcfs.takeoff(targetHeight=Z, duration=2.0)

    timeHelper.sleep(4)
    
    cf_one.goTo(cf_two.position(), 0, 6)
    cf_two.goTo(cf_one.position(), 0, 6)

    timeHelper.sleep(7)

    allcfs.land(targetHeight=0.06, duration=2.0)

    timeHelper.sleep(3)


if __name__ == "__main__":
    main()
```

**Comment by sAz-G** [comment]
Hi,

I am having the same problem. I am also using the V1 system and the same Crazyflies. However, I just tried to run the provided script [collisionAvoidance.py](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/collisionAvoidance.py) and it did not work. The drones fly, but it does not look like they are avoiding each other.

I followed the installation instructions on [this website](https://crazyswarm.readthedocs.io/en/latest/index.html) and configured the system accordingly.

**Comment by jpreiss** [comment]
for collision avoidance to occur, the CFs must be on the same radio channel.

**Reply by sAz-G** [reply]
In my case, I set the channel to 80 (instead  of the default 100). The drones are able to communicate with the pc (they receive commands, and send back information). Does the collision avoidance script use another channel? Is the information of the collision avoidance being sent on a different channel?

**Reply by jpreiss** [reply]
They must all be on the same channel, but it doesn't matter which channel.

Each CF's position is broadcasted to anyone listening on the radio channel. Collision avoidance knows the position of other CFs by listening to all messages on the broadcast channel.

**Comment by Mardank** [comment]
For my side, both CF are on the same channel (ID radio://0/100/2M/E7E7E7E700 and radio://0/100/2M/E7E7E7E701), but there is still no sign of any collision avoidance when running the program above.

Is there anything else that should be taken into account? Could there be a way to figure out what's going wrong?

**Reply by jpreiss** [reply]
Maybe you are using an old firmware?

**Reply by Mardank** [reply]
I flashed the drones using the cfclient with the latest firmware version 2022.05 and even tried the firmware version you tested (2021.06).

**Reply by jpreiss** [reply]
No other obvious ideas unfortunately. You can try making the parameters of the collision avoidance more conservative (see bottom of src/modules/src/collision_acoidance.c)

**Comment by whoenig** [comment, marked as answer]
Lighthouse is a distributed localization system, so each drone just estimates its own position (and by default keeps it to itself). The collision-avoidance mode we have assumes that each drone knows about the neighboring drones. When using a motion capture, this knowledge is easy to obtain, because Crazyswarm broadcasts the position information of *all* Crazyflies (so every single one knows where all members of the swarm are). In case of the LightHouse, you would need to implement this information sharing yourself using the peer2peer communication functionality.

**Reply by sAz-G** [reply]
Is it possible to let the pc log the position from each drone, and then broadcast to everyone? Or is it the same as using p2p?

Basically I am trying to understand if logging the position from each drone is faster than a p2p round (if round robin p2p is implemented), and if there are constraints on logging that I have to consider.

**Reply by whoenig** [reply]
Logging is more bandwidth constraint (you have more messages in that case). Using P2P is better. In the P2P system, each drone sends actually broadcasts, so no need to do any round robin scheduling.
